src/types.ts

```diff
@@ -0,0 +1,215 @@
+export type Unparsed = unknown
+
+export type Uint = (number | bigint) & { __type: 'uint' }
+export type Int = (number | bigint) & { __type: 'int' }
+
+export type MaxlenString<N> = string & { __maxLength: N }
+export type FixlenBuffer<N> = Buffer & { __length: N }
+export type MaxlenBuffer<N> = Buffer & { __maxLength: N }
+
+export const ADDRESS_KEY_HASH_LENGTH = 28
+export const SCRIPT_HASH_LENGTH = 28
+export const GENESIS_DELEGATE_HASH_LENGTH = 28
+export const POOL_KEY_HASH_LENGTH = 28
+export const GENESIS_HASH_LENGTH = 28
+export const VRF_KEY_HASH_LENGTH = 32
+export const METADATA_HASH_LENGTH = 32
+export const TX_ID_HASH_LENGTH = 32
+
+export const IPV4_LENGTH = 4
+export const IPV6_LENGTH = 16
+
+export const ASSET_NAME_MAX_LENGTH = 32
+export const URL_MAX_LENGTH = 64
+export const DNS_NAME_MAX_LENGTH = 64
+
+export type Address = Buffer
+export type RewardAccount = Buffer
+export type Coin = Uint
+export type Epoch = Uint
+
+// Multiasset
+export type PolicyId = FixlenBuffer<typeof SCRIPT_HASH_LENGTH>
+export type AssetName = MaxlenBuffer<typeof ASSET_NAME_MAX_LENGTH>
+export type Multiasset<T> = Map<PolicyId, Map<AssetName, T>>
+
+// Transaction input
+export type TransactionInput = {
+    transactionId: FixlenBuffer<typeof TX_ID_HASH_LENGTH>,
+    index: Uint,
+}
+
+// Transaction output
+export enum AmountType {
+    WITHOUT_MULTIASSET,
+    WITH_MULTIASSET,
+}
+
+export type Amount = {
+    type: AmountType.WITHOUT_MULTIASSET,
+    coin: Coin,
+} | {
+    type: AmountType.WITH_MULTIASSET,
+    coin: Coin,
+    multiasset: Multiasset<Uint>,
+}
+
+export type TransactionOutput = {
+    address: Address,
+    amount: Amount,
+}
+
+// Certificate
+export enum CertificateType {
+    STAKE_REGISTRATION = 0,
+    STAKE_DEREGISTRATION = 1,
+    STAKE_DELEGATION = 2,
+    POOL_REGISTRATION = 3,
+    POOL_RETIREMENT = 4,
+    GENESIS_KEY_DELEGATION = 5,
+    MOVE_INSTANTANEOUS_REWARDS_CERT = 6,
+}
+
+export enum StakeCredentialType {
+    ADDRESS_KEY_HASH = 0,
+    SCRIPT_HASH = 1,
+}
+
+export type StakeCredentialAddr = {
+    type: StakeCredentialType.ADDRESS_KEY_HASH,
+    hash: FixlenBuffer<typeof ADDRESS_KEY_HASH_LENGTH>,
+}
+
+export type StakeCredentialScript = {
+    type: StakeCredentialType.SCRIPT_HASH,
+    hash: FixlenBuffer<typeof SCRIPT_HASH_LENGTH>,
+}
+
+export type StakeCredential = StakeCredentialAddr | StakeCredentialScript
+
+export type StakeRegistrationCertificate = {
+    type: CertificateType.STAKE_REGISTRATION,
+    stakeCredential: StakeCredential,
+}
+
+export type StakeDeregistrationCertificate = {
+    type: CertificateType.STAKE_DEREGISTRATION,
+    stakeCredential: StakeCredential,
+}
+
+export type StakeDelegationCertificate = {
+    type: CertificateType.STAKE_DELEGATION,
+    stakeCredential: StakeCredential,
+    poolKeyHash: FixlenBuffer<typeof POOL_KEY_HASH_LENGTH>,
+}
+
+export type UnitInterval = [Uint, Uint]
+
+export type Port = Uint & { __le: 65535 }
```

Comment:
```suggestion
export type Port = Uint & { __maxSize: 65535 }
```
I would also consider making this a branded type similar to `MaxlenString`. Perhaps `MaxsizeUint`? This can then also be used in the `isUintLe` function.